libvast/src/system/configuration.cpp

```diff
@@ -79,32 +83,99 @@ caf::error configuration::parse(int argc, char** argv) {
   command_line.erase(caf_opt, command_line.end());
-  for (auto& arg : caf_args) {
-    // Remove caf. prefix for CAF parser.
-    if (detail::starts_with(arg, "--caf."))
-      arg.erase(2, 4);
-    // Rewrite --config= option to CAF's expexted format.
+  // If the user provided a config file on the command line, we attempt to
+  // parse it last.
+  for (auto& arg : command_line)
     if (detail::starts_with(arg, "--config="))
-      arg.replace(8, 0, "-file");
+      config_paths.push_back(arg.substr(9));
+  // Parse and merge all configuration files.
+  caf::settings merged_settings;
+  for (const auto& config : config_paths) {
+    if (exists(config)) {
+      auto contents = load_contents(config);
+      if (!contents)
+        return contents.error();
+      auto yaml = from_yaml(*contents);
+      if (!yaml)
+        return yaml.error();
+      auto rec = caf::get_if<record>(&*yaml);
+      if (!rec)
+        return caf::make_error(ec::parse_error, "config file not a YAML map");
+      auto flat_yaml = flatten(*rec);
+      // Erase all null values because a caf::config_value has no such notion.
+      for (auto i = flat_yaml.begin(); i != flat_yaml.end();) {
+        if (caf::holds_alternative<caf::none_t>(i->second))
+          i = flat_yaml.erase(i);
+        else
+          ++i;
+      }
+      auto settings = to<caf::settings>(flat_yaml);
+      if (!settings)
+        return settings.error();
+      detail::merge_settings(*settings, merged_settings);
+    }
   }
-  for (const auto& p : config_paths) {
-    if (auto err = actor_system_config::parse({}, p.str().c_str())) {
-      err.context() += caf::make_message(p);
-      return err;
+  // TODO: Revisit this after we are on CAF 0.18.
+  // Helper function to parse a config_value with the type information
+  // contained in an config_option. Because our YAML config only knows about
+  // strings, but a config_option may require an atom, we have to use a
+  // heuristic to see whether either type works.
+  auto parse_config_value
+    = [](const caf::config_option& opt,
+         const caf::config_value val) -> caf::expected<caf::config_value> {
+    // Hackish way to get a string representation that doesn't add double
+    // quotes around the value.
+    auto no_quote_stringify
+      = detail::overload([](const auto& x) { return caf::deep_to_string(x); },
+                         [](const std::string& x) { return x; });
+    auto str = caf::visit(no_quote_stringify, val);
+    auto result = opt.parse(str);
+    if (!result) {
+      // We now try to parse strings as atom using a regex, since we get
+      // recursive types like lists for free this way. A string-vs-atom type
+      // clash is the only instance we currently cannot distinguish. Everything
+      // else is a true type clash.
+      // (With CAF 0.18, this heuristic will be obsolete.)
+      str = std::regex_replace(str, std::regex("\""), "'");
```

Comment:
Please rewrite this in terms of `std::transform_if` for three reasons:
1. It is absolutely overcomplicating the code for readers and reviewers.
2. It is unnecessarily slow at runtime, especially the unnecessarily repeated regular expression compilation (these should _always_ be cached).
3. The inclusion of the regex header adds a lot of compile time overhead.

Reply:
Turns out we had `detail::replace_all` already! 🙈